### DISCUSSION TITLE
libmesh error - You requested to reuse the nonlinear residual vector as the base vector

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#19496] [Q&A General] [answered]
I ran into this error that seems to trace back to libmesh, i was wondering if someone could give me some insight. 

`You requested to reuse the nonlinear residual vector as the base vector for computing the action of the matrix-free Jacobian, but the vectors are not the same. Your physics must have states; either remove the states from your code or make sure that you set_mf_reuse_base(false)`

Thanks,
Zack

**Comment by GiudGiud** [comment]
@fdkong

**Comment by fdkong** [comment, marked as answer]
@rtaylo45 Could you briefly state what problems are you solving? Are you working on contact problems? Does your problem have discontinuities?

**Reply by rtaylo45** [reply]
The issue kinda comes up in a weird way. I am coupling SAM and an ORNL moose based app called Mole. I am compiling the codes together in the Mole make file. I can break this problem down into two cases:

1. Clone down both repos, i go into the SAM repo and call make. Go into the Mole repo call make. Run my unit tests and get some exodiff errors with the test associated with coupled Mole/SAM test (not the topic of this discussion) . I run a problem that includes a SAM flow loop and everything runs fine.
2. Clone down both repos, i go into the Mole repo and call make. The make file tells the code to go and compile SAM and link to it. Run my unit tests, which all pass. But when i try to run the flow loop problem, i get the error i showed above. 

What's interesting is that the problem only seems to show up for flow loop problems in SAM. Its possible that it will show up in other cases, but i have not found a case where it has. Im not exactly sure what SAM does under the hood to solve these flow problems but i can reach out to them and ask.

**Reply by fdkong** [reply]
It is interesting. There are two modes in SAM: "stand-alone" and "work-with-other." That might explain the weird behavior you saw here. Option 1 you did should involve the SAM stand-alone mode that SAM developers might test extensively.  The option 2 you tried was the "work-with-other" mode, which might not be fully explored. 

It could be another reason as well. We have an option to turn this error message off "snesmf_reuse_base=false" (executioner block), but we need to get to the bottom of the problem instead of just turning off messages.  

What was happening was that PETSc saw the nonlinear function change from nonlinear and linear iterations. That is not common, and in moose, that occurs primarily in mechanics contact module because of contact.  

A SAM developer might be helpful here @lingzou

**Reply by travismui** [reply]
Hi @fdkong @rtaylo45,
I believe in SAM we actually set this param `snesmf_reuse_base = false` by default for our standalone mode. In the coupled Mole/SAM, this option is probably not being set which is why you're seeing the error messages. We are not modeling mechanical contact but there may be other factors that prevent us from using this optimization in some scenarios.

**Reply by GiudGiud** [reply]
Turning terms on or off can create this error right @fdkong ?
I think this is a common thing to do for 1D TH.

**Reply by fdkong** [reply]
If the function has a state, we can not reuse the base from the nonlinear iteration. In this case, we need to use "snesmf_reuse_base = false" to disable that, otherwise you will see an error message. By default, we have "snesmf_reuse_base = true" in moose, which means we want to reuse the base from the nonlinear iteration to reduce the number of residual evaluations that is not cheap.

I am not sure whether or not 1D TH has a state. The equations look just fine to me. There are some special materials?

If we are sure that the function has a state, then it is safe to turn the message off. Otherwise that could be a bug somewhere

**Comment by sapitts** [comment]
I've hit this same error in creating a regression test for a new tensor mechanics (crystal plasticity), no contact. I was able to get around the error in opt by moving all of my stateful increment calculations into a single method, but then hit the same error as given in the original posting in dbg.

Clearly I'm piggy-backing on this discussion topic, yet are there any recommendations for how to write materials to avoid this error? I've got an existing PR with the failing regression test and problematic code up ( #19502 ) if it peaks anyone's interest

**Reply by aarograh** [reply]
@sapitts did you ever get any resolution on this?  I am currently running into this error as well in trying to update an app's MOOSE from March 28 to a more recent commit (July 20).

**Reply by sapitts** [reply]
Hi @aarograh, I ended up lagging some of the stateful material properties, that were used to calculated other stateful material properties, in order to avoid the error and error message. There's a good chance that the error worked as intended and helped me avoid conflicting states. I hope that you can resolve the error in your code too!

**Reply by hugary1995** [reply]
I would second Fande's suggestion above: if your residual has state, then you can just set snesmf_reuse_base to false. If your residual doesn't have state, this is a bug. You should get to the bottom of this, otherwise you may get nondeterministic results.

**Reply by aarograh** [reply]
I'm getting back into MOOSE after a couple years out of it, so I'm probably going to ask some stupid questions here.  What do you mean by "if your residual has state"?  This particular failure is for a test case of a pretty simple kernel in our application.  The fact that it works with MOOSE from March but not July is pretty concerning.

I tried setting the `snesmf_reuse_base = false` and the solve just completely fails to converge the first timestep.  So based on your comment, there is a real problem going on here, but I'm not clear on where to begin in searching for the issue.  Perhaps a better understanding of "state"/"stateful" would clarify things for me.

**Reply by sapitts** [reply]
We use here the phrase 'residual has state' to indicate that material properties, which depend on `old` material properties, from the previous timestep, are used in the calculation of the weak form of the PDE. 

As an example, in rate-dependent tensor mechanics models, we calculate the stress divergence as a function of material properties (e.g. yield surface) that depend on the previous value (from the previous timestep) of that yield surface. We would say this yield surface material property is stateful or has state. This yield stress is used to calculate the stress, and that stress is used to calculate the stress divergence. In a tensor mechanics FEM simulation, we drive the stress divergence to as close to zero as we can (or within a tolerance), and that difference between the ideal zero and actual FEM calculation value is the residual. Because the residual depends on a stateful material property, through this chain, we say here that the residual has state. Hope this helps

If you have multiple stateful material properties in your code, as I did, I wonder if there is a potential for those stateful properties to get out of sync. If it was me debugging this type of problem, I'd start by checking to make sure all the material properties were initialized as expected and then walk through the system to monitor the values. I'm interested to hear the debugging recommendations from @hugary1995 too!
